### PR TITLE
[AMDGPU] Optimize out s_barrier_signal/_wait

### PR DESCRIPTION
**Pull request by piotrAMD**
Extend the optimization that converts s_barrier to wave_barrier (nop)
when the number of work items is not larger than wave size.
    
This handles the "split barrier" form of s_barrier where the barrier
is represented by separate intrinsics (s_barrier_signal/s_barrier_wait).
Note: the version where s_barrier is used in gfx12 (and later split)
has the optimization already, but some front-ends may prefer to use
split intrinsics and this is being addressed by the patch.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll [hunk 1]
```diff
@@ -0,0 +1,56 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1200 < %s | FileCheck %s
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1200 < %s -global-isel | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll:2]: move -global-isel to the front and use explicit =0 for the dag checks

File: llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+; CHECK: s_barrier_wait
+define amdgpu_kernel void @wait_unknown_wgs() {
+  tail call void @llvm.amdgcn.s.barrier.wait(i16 -1) #0
+  ret void
+}
+
+; CHECK-LABEL: {{^}}wait_flat_wgs_attr_32_128:
+; CHECK: s_barrier_wait
+define amdgpu_kernel void @wait_flat_wgs_attr_32_128() #1 {
+  tail call void @llvm.amdgcn.s.barrier.wait(i16 -1) #0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll:38]: Don't need the callsite attributes

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 3]
```diff
@@ -9605,7 +9605,9 @@ SDValue SITargetLowering::LowerINTRINSIC_VOID(SDValue Op,
     unsigned Opc = Done->isZero() ? AMDGPU::EXP : AMDGPU::EXP_DONE;
     return SDValue(DAG.getMachineNode(Opc, DL, Op->getVTList(), Ops), 0);
   }
-  case Intrinsic::amdgcn_s_barrier: {
+  case Intrinsic::amdgcn_s_barrier:
+  case Intrinsic::amdgcn_s_barrier_signal:
```
Comment (perlfu) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:9609]: This is lowering both the signal and the wait to `WAVE_BARRIER`, but I wonder if really the signal should just be removed entirely?

File: llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll (hunk 1 above)
Comment (piotrAMD) [reply llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll]: Done.

File: llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll (hunk 2 above)
Comment (piotrAMD) [reply llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll]: Done.

File: llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+; CHECK-LABEL: {{^}}signal_flat_wgs_attr_32_128:
+; CHECK: s_barrier_signal
+define amdgpu_kernel void @signal_flat_wgs_attr_32_128() #1 {
+  tail call void @llvm.amdgcn.s.barrier.signal(i32 -1)
+  ret void
+}
+
+; CHECK-LABEL: {{^}}signal_flat_wgs_attr_16_32:
+; CHECK: :
+; CHECK-NEXT: s_endpgm
```
Comment (rovka) [inline llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll:20]: Nit: Would it be more robust to `CHECK-NOT: barrier` instead?

File: llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll [hunk 5]
```diff
@@ -0,0 +1,55 @@
+; RUN: llc -global-isel=0 -mtriple=amdgcn -mcpu=gfx1200 < %s | FileCheck %s
```
Comment (shiltian) [inline llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll:1]: I strongly suggest to auto gen the check lines please, which can make other's life much easier in the future.

Comment (piotrAMD) [reply llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll:2]: Done.

File: llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll (hunk 4 above)
Comment (piotrAMD) [reply llvm/test/CodeGen/AMDGPU/barrier-elimination-gfx12.ll]: Good point - regenerated check lines with update_llc_test_checks.py to make it more robust.

**Review by ruiling — APPROVED** [review]
LGTM

**Pull request merged**